### ISSUE TITLE
SteamVR 1.9 Breaks `VROverlayFlags_NoDashboardTab`

### ISSUE THREAD

**Issue by Kyrluckechuck** [ValveSoftware/openvr#1270]
Disclaimer:
Tested using the HTC Vive and HTC Vive Pro using the latest openvr API release (1.8.19) on a Windows 10 Pro 64-bit system (Build 18363).

Since the public release of SteamVR 1.9, we've now realized our old method of hiding the built-in overlay tabs using the `overlay->SetOverlayFlag(tabHandle, vr::VROverlayFlags::VROverlayFlags_NoDashboardTab, true)` method, no longer works. 

Below is a screenshot showing which tabs I am referring to in case that is in question:
![vrmonitor_gJRW9HUnJw](https://user-images.githubusercontent.com/7606153/71100410-f8b07380-2182-11ea-987c-53df7d9e85e7.png)

Using `overlay_viewer.exe` from the SteamVR bin, I located a couple new dashboard tabs and that a couple had changed names, which isn't a huge deal so long as we're able to find them. However, after setting the code to then hide those overlays, it appears that we are no longer able to affect them *at all*. 

The base code we use is below, scraped together from various sources over the past few years, though right now it has some extra attempts at trying to hide it such as the pointer, thumb, and destroy lines. This has been very reliable since first implemented way back, with our overlay even being able to re-show any tabs when our overlay closes (ideal for testing and other uses). 

Any more details I can provide, please just let me know.

```cpp
void OverlayController::HideTabs() {
    qDebug() << "Hiding tabs";
//    HideTab("valve.steam.desktop"); // No longer exists
    HideTab("system.desktop.1");
    HideTab("system.desktop.1.pointer"); // Even attempting Pointer & Thumb in case
    HideTab("system.desktop.1.thumb");
    HideTab("system.desktop.2");
    HideTab("valve.steam.bigpicture");
//    HideTab("htc_vive_overlay");  // No longer exists
    HideTab("system.settings");
//    HideTab("system.vrdashboard"); // No longer exists
}

vr::VROverlayHandle_t OverlayController::HideTab(QString key) {
    vr::IVROverlay* overlay = vr::VROverlay();

    if (overlay == nullptr)
        return 0;

    ULONGLONG tabHandle = 0;
    vr::EVROverlayError findOverlayError = vr::EVROverlayError::VROverlayError_UnknownOverlay;

    for(int i = 0; ++i < 6;) {
        findOverlayError = overlay->FindOverlay(key.toLocal8Bit().data(), &tabHandle);

        if (findOverlayError != vr::EVROverlayError::VROverlayError_None) {
           Sleep(250);
        }
    }

    if (findOverlayError != vr::EVROverlayError::VROverlayError_None) {
        qDebug() << "Couldn't find overlay '" << key << "': " << findOverlayError;
        return 0;
    }
    qDebug() << "Disabling " << key;
    overlay->SetOverlayAlpha(tabHandle, 0);  // Hides only the actual view pane of the element
    overlay->SetOverlayFlag(tabHandle, vr::VROverlayFlags::VROverlayFlags_NoDashboardTab, true); // The old, reliable method to hide the tab
    overlay->SetOverlayFlag(tabHandle, vr::VROverlayFlags::VROverlayFlags_VisibleInDashboard, false); // An attempt to hide it using any method exposed
    overlay->HideOverlay(tabHandle); // No effect but figured worth trying
    overlay->DestroyOverlay(tabHandle); // No effect but figured worth trying
    qDebug() << "Disabled " << key;
    return tabHandle;
}
```

**Comment by RaymondSmit** [comment]
Just wondering: Are you hiding overlays that are not yours? I'm the developer of an app that uses overlays (Stop Sign VR), but I wouldn't like other apps closing or hiding mine. 
(If the user wants to close it, that's fine of course.)

**Comment by Kyrluckechuck** [comment]
Totally fair point @RaymondSmit !

In most situations, that would make sense, I completely agree.

Our specific use case is different from most, though, as this software only runs on closed-systems in which the customer is solely interacting with the environment we provide, not with any other software (and they cannot install or run any other software). We provide the entire experience and try to remove any elements in which the consumer would be potentially confused by it -- since we can configure/reconfigure the systems ourselves, we don't need to allow them access to things such as the `Desktop 1` or `Settings` views, as they only cause issues or confusion.

**Comment by RaymondSmit** [comment]
OK, got it. As in a VR arcade setting.
Maybe you can start SteamVR without those unwanted tabs?

**Comment by Kyrluckechuck** [comment]
@RaymondSmit correct, exactly that, yeah! Much simpler way to describe it.

It would be ideal, in general, to be able to control the tabs visible, but are you saying it's possible to launch SteamVR without the tabs at all without editing the HTML/CSS? I have heard this referenced a few times across the web, usually with people asking for the feature, but have yet to come across any documentation of such a feature existing -- this would definitely work for us (though I'd still leave this issue open so that it is documented about not working)!

**Comment by bddckr** [comment]
I have the same needs, yet editing files from the SteamVR installation wouldn't be a solution for us at all. We just don't want to adjust third party content at all.

The existing APIs did a good job. If a user is not liking the fact that overlays (or their tabs) are hidden this brings up a case of the user telling the author of said "app" to stop doing it or offer configuration. They can stop using the third-party app until then.

Please do not break/remove even more available OpenVR API.

**Comment by RaymondSmit** [comment]
Starting SteamVR with some command-line switches to not show those tabs may be a good solution.
I don't know if that's possible already.

@bddckr Sure, I agree, when a user does not want an app running, he should be able to close it. But, IMO, other apps shouldn't be able to stop any apps or overlays they don't own or didn't start.

**Comment by bddckr** [comment]
I think they should if that's the goal of the user. This isn't currently an issue, otherwise we'd hear from users. Making it an artificial issue for the sake of just restricting existing API for no benefit hurts the platform IMO. Other platforms have solved this issue (once it came up) by introducing a permission system, giving the user the full ability to control what's possible and what's not.

**Comment by RaymondSmit** [comment]
Not all users would understand why they don't see some app or overlay anymore.
The max number of overlays was increased to 128 recently. What if an app uses many overlays and reaches the limit? It may decide to close some overlays of other apps.

**Comment by Kyrluckechuck** [comment]
@RaymondSmit to clarify, the issue described hear is not trying to destroy other overlays (though I did try to just to see if it would change the effect). 

The exact issue here is the inability to hide the tabs of other overlays (such as Settings or Desktop 1), which up until now has not been an issue. The overlays still persist and are only not showing their tabs, and can be re-enabled to show them again. If any given app is scared it might have this happen to it and *really* doesn't want to be hidden, they can release themselves among other solutions.

As @bddckr raised though, a permissions system in the future might be a good solution to avoiding abuse if it ever becomes an issue, but as of right now it's just a feature that has been (seemingly) unintentionally broken.

**Comment by JoeLudwig** [comment]
VROverlayFlags_NoDashboardTab is used by the controller binding UI internally, so that flag is definitely working.

The way dashboard tabs are collected and displayed changed significantly in 1.9. I think what might have happened in this case is that setting the flag well after the dashboard tab has been rendered doesn't work.

I assume that the black square in your screenshot is your own overlay, and that what you want to turn off is:  
* settings
* Steam
* Desktop

There is a user setting to turn off the in-headset settings (for exactly this kind of scenario.) It's called "Access Settings In-Headset and is an advanced option in the General section.

Do you need Steam to be running while you run SteamVR on these machines? Not running Steam would get rid of the Steam overlay.

Which leaves desktop... Given the purpose of the setting-hiding-setting, it makes sense to also include the desktop in that setting (and maybe rename it.) I'll run that up the flagpole around here.

I think using settings for this makes more sense than requiring an API call in any case. You can include an "override.vrsettings" file in your install config directory and the settings in that file will win out over any local settings on that machine. It essentially disables the settings UI and APIs everywhere.

**Comment by Kyrluckechuck** [comment]
@JoeLudwig Thanks for the speedy reply!

That would make more sense, since it seems that every now and then the tabs will disappear, but it seems to be very random and not reproducible (which would make sense if it has to do with the rendering order).

You are correct, I had thought it non-relevant to keep the icon there, it's just another overlay app on the bar - so to speak.

Correct, the Settings would not be an issue to turn off via the General section, and would be quite ideal if turning that off also turned off the Desktop tab(s), too.

Unfortunately we do use Steam and so it would be active on the computers at the same time, whether it be for launching SteamVR with the latest updates to launching some games/apps using the steam protocols, so that won't resolve it 😟

I'm totally on board with the `override.vrsettings` idea, is there any documentation about what that JSON would include, such as how to list the tabs that should be enabled/disabled, for example?

**Comment by JoeLudwig** [comment]
We fixed setting this flag on overlays in 1.9.14: https://steamcommunity.com/games/250820/announcements/detail/1699477947720577647
